### PR TITLE
Expose RemotePointEvaluation parameters in non-nested multigrid

### PR DESCRIPTION
**Pull request by fdrmrc**
...through an AdditionalData struct.

### CODE REVIEW

**Comment by fdrmrc** [comment]
In addition to the arguments needed by the constructor of RemotePointEvaluation, I've added an additional boolean, `enfroce_all_points_found`, that allows to throw if not all points have been found after the search procedure. Ultimately, it calls the `RPE::all_points_found()`. 

To give more context, this PR is motivated by the fact that some boundary points may not be picked up due to floating point arithmetic in certain situations (for instance when one has neumann boundary conditions), leading to dramatic effects on the number of required iterations. This issue can be solved by simply tweaking the tolerance parameter, and this is what the present PR allows.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h [hunk 1]
```diff
@@ -706,6 +706,35 @@ class MGTwoLevelTransferNonNested<dim,
...
+   * AdditionalData structure with the arguments needed by
+   * RemotePointEvaluation. Default values are the same as the ones described in
+   * the documentation of RemotePointEvaluation. The last boolean parameter, @p enf_all_points_found is true by defaults and
+   * checks if RemotePointEvaluation::all_points_found() evaluates to true, i.e.
+   * all submitted points have been found inside the domain.
+   */
+  struct AdditionalData
+  {
+    AdditionalData(const double       tol                = 1e-6,
+                   const bool         enf_unique_mapping = false,
```
Comment (peterrum) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:719]: Not needed.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h [hunk 2]
```diff
@@ -706,6 +706,35 @@ class MGTwoLevelTransferNonNested<dim,
...
+                   const unsigned int rtree_l            = 0,
+                   const std::function<std::vector<bool>()> &marked_verts = {},
```
Comment (peterrum) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:721]: Not needed.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h [hunk 3]
```diff
@@ -706,6 +706,35 @@ class MGTwoLevelTransferNonNested<dim,
...
+                   const bool enf_all_points_found = true)
```
Comment (peterrum) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:722]:
```suggestion
                   const bool enforce_all_points_found = true)
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h [hunk 4]
```diff
@@ -788,7 +818,7 @@ class MGTwoLevelTransferNonNested<dim,
    * Object to evaluate shape functions on one mesh on visited support points of
    * the other mesh.
    */
-  Utilities::MPI::RemotePointEvaluation<dim> rpe;
+  std::shared_ptr<Utilities::MPI::RemotePointEvaluation<dim>> rpe;
```
Comment (peterrum) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:821]: The change to shared_ptr is not needed!?

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 5]
```diff
@@ -4170,13 +4183,19 @@ MGTwoLevelTransferNonNested<dim, LinearAlgebra::distributed::Vector<Number>>::
       this->partitioner_fine->global_to_local(global_dof_indices[i]);
 
   // hand points over to RPE
-  rpe.reinit(points, dof_handler_coarse.get_triangulation(), mapping_coarse);
+  rpe->reinit(points, dof_handler_coarse.get_triangulation(), mapping_coarse);
+
+  if (additional_data.enforce_all_points_found)
+    AssertThrow(
+      rpe->all_points_found(),
```
Comment (peterrum) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:4190]:
```suggestion
      !additional_data.enforce_all_points_found || rpe->all_points_found(),
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h (hunk 4 above)
Comment (fdrmrc) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.h]: Sure, it was a leftover from debugging. Fixed, thanks

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h [hunk 6]
```diff
@@ -706,6 +706,32 @@ class MGTwoLevelTransferNonNested<dim,
...
 public:
+  /**
+   * AdditionalData structure that can be used to tweak parameters
+   * @p tolerance and @p rtree_level related to the search procedure (used
+   * internally by RemotePointEvaluation) or, in the future, transfer operators
+   * needed by the non-nested multigrid algorithm. Default values are the same
+   * as the ones described in the documentation of RemotePointEvaluation.
+   * The last Boolean parameter, @p enforce_all_points_found is true by default
+   * and checks if RemotePointEvaluation::all_points_found() evaluates to true,
+   * i.e. all submitted points have been found inside the domain.
```
Comment (peterrum) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:717]: Normally, we described these parameters closer to the parameters. See https://github.com/dealii/dealii/blob/dfb435053092192fda1ca9a76a1597e589155275/include/deal.II/lac/solver_gmres.h#L197-L261.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h [hunk 7]
```diff
@@ -4172,6 +4180,11 @@ MGTwoLevelTransferNonNested<dim, LinearAlgebra::distributed::Vector<Number>>::
   // hand points over to RPE
   rpe.reinit(points, dof_handler_coarse.get_triangulation(), mapping_coarse);
 
+  AssertThrow(
+    !additional_data.enforce_all_points_found || rpe.all_points_found(),
+    ExcMessage(
+      "You requested that all points should be found, but this didn't happen. You can change this option through the AdditionaData struct in the constructor."));
```
Comment (peterrum) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:4186]: Could you break the line?

**Review by peterrum — APPROVED** [review]
Looks good. Just some minor documentation changes.

**Review by peterrum — APPROVED** [review]
Thanks!

**Pull request merged**